stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/adapter/view/FileAttachmentsView.kt

```diff
@@ -121,20 +136,60 @@ private class FileAttachmentViewHolder(
         binding.apply {
             fileTypeIcon.loadAttachmentThumb(attachment)
             fileTitle.text = attachment.getDisplayableName()
-            fileSize.text = MediaStringUtil.convertFileSizeByteCount(attachment.fileSize.toLong())
 
-            if (attachment.uploadComplete == true || attachment.uploadComplete == null) {
-                actionButton.setImageResource(R.drawable.stream_ui_ic_icon_download)
-                val tintColor = ContextCompat.getColor(context, R.color.stream_ui_black)
-                ImageViewCompat.setImageTintList(actionButton, ColorStateList.valueOf(tintColor))
-            } else if (attachment.uploadComplete == false) {
+            if (attachment.uploadState is Attachment.UploadState.Failed) {
                 actionButton.setImageResource(R.drawable.stream_ui_ic_warning)
                 val tintColor = ContextCompat.getColor(context, R.color.stream_ui_accent_red)
                 ImageViewCompat.setImageTintList(actionButton, ColorStateList.valueOf(tintColor))
+                fileSize.text = MediaStringUtil.convertFileSizeByteCount(attachment.upload?.length() ?: 0L)
+            } else {
+                actionButton.setImageResource(R.drawable.stream_ui_ic_icon_download)
+                val tintColor = ContextCompat.getColor(context, R.color.stream_ui_black)
+                ImageViewCompat.setImageTintList(actionButton, ColorStateList.valueOf(tintColor))
+                fileSize.text = MediaStringUtil.convertFileSizeByteCount(attachment.fileSize.toLong())
+            }
+
+            binding.progressBar.isVisible = attachment.uploadState is Attachment.UploadState.InProgress
+
+            if (attachment.uploadState is Attachment.UploadState.InProgress) {
+                attachment.uploadId?.let(ProgressTrackerFactory::getOrCreate)?.let { tracker ->
+                    val progress = tracker.currentProgress()
+                    val completion = tracker.isComplete()
+                    val totalValue = MediaStringUtil.convertFileSizeByteCount(
+                        attachment.upload?.length() ?: 0
+                    )
+
+                    val fileProgress = progress.combine(completion, ::Pair)
+
+                    clearScope()
+                    scope = CoroutineScope(DispatcherProvider.Main)
+
+                    scope!!.launch {
+                        fileProgress.collect { (progress, isComplete) ->
+                            if (!isComplete) {
```

Comment:
Please extract this to some smaller method. It's very hard to read (something like `handleUploadResult` )